include/std/experimental/__detail/easy_cancel.hpp

```diff
@@ -0,0 +1,68 @@
+// Copyright (c) 2023 Gor Nishanov
+// Licensed under MIT license. See LICENSE.txt for details.
+
+#ifndef _STD_EXPERIMENTAL_CONQUEUE_EASY_CANCEL
+#define _STD_EXPERIMENTAL_CONQUEUE_EASY_CANCEL
+
+#include <cassert>
+#include <stdexec/execution.hpp>
+#include <variant>
+
+namespace std::experimental::__detail {
+
+template <typename Receiver, typename CancelCallback, typename StopToken,
+          bool Unstoppable>
+struct easy_cancel_base;
+
+template <typename Receiver, typename CancelCallback, typename StopToken>
+struct easy_cancel_base<Receiver, CancelCallback, StopToken, false> {
+  using StopCallback =
+      typename StopToken::template callback_type<CancelCallback>;
+
+  variant<monostate, StopToken, StopCallback> state;
+
+  easy_cancel_base(Receiver& receiver)
+      : state(stdexec::get_stop_token(stdexec::get_env(receiver))) {}
+
+  bool stop_requested() {
+    if (auto* tok = std::get_if<StopToken>(&state))
+      return tok->stop_requested();
+
+    return false;
+  }
+
+  void register_callback(CancelCallback&& f) {
+    if (auto* tok = std::get_if<StopToken>(&state))
+      state.template emplace<2>(std::move(*tok), std::move(f));
```

Comment:
I am not confident that the token is used before it is destroyed. I make a copy of the value before the emplace when I use this idiom.

Reply:
Got it. Move token out, then emplace!

Reply:
I double check. variant indeed, first destroy the existing element, then move constructs. :-)
Doing this as a one line was busted :-)

```c++
    variant<monostate, A, B> v;
    puts("started");
    v.emplace<1>('a');
    v.emplace<2>(std::move(get<1>(v)));
    puts("ended");
```
output:
```
started
construct 0x404051 with a
destruct 0x404051 with a
move construct 0x404051 from 0x404051 with a
ended
destruct 0x404051 with a
```